doc/man1/qgpumode.1.in
.\"     OpenPBS (Portable Batch System) v2.3 Software License
.\" 
.\" Copyright (c) 1999-2000 Veridian Information Solutions, Inc.
.\" All rights reserved.
.\" 
.\" ---------------------------------------------------------------------------
.\" For a license to use or redistribute the OpenPBS software under conditions
.\" other than those described below, or to purchase support for this software,
.\" please contact Veridian Systems, PBS Products Department ("Licensor") at:
.\" 
.\"  www.OpenPBS.org +1 650 967-4675         sales@OpenPBS.org
.\"            877 902-4PBS (US toll-free)
.\" ---------------------------------------------------------------------------
.\" 
.\" This license covers use of the OpenPBS v2.3 software (the "Software") at
.\" your site or location, and, for certain users, redistribution of the
.\" Software to other sites and locations. Use and redistribution of
.\" OpenPBS v2.3 in source and binary forms, with or without modification,
.\" are permitted provided that all of the following conditions are met.
.\" After December 31, 2001, only conditions 3-6 must be met:
.\" 
.\" 1. Commercial and/or non-commercial use of the Software is permitted
.\"  provided a current software registration is on file at www.OpenPBS.org.
.\"  If use of this software contributes to a publication, product, or service
.\"  proper attribution must be given; see www.OpenPBS.org/credit.html
.\" 
.\" 2. Redistribution in any form is only permitted for non-commercial,
.\"  non-profit purposes. There can be no charge for the Software or any
.\"  software incorporating the Software. Further, there can be no
.\"  expectation of revenue generated as a consequence of redistributing
.\"  the Software.
.\" 
.\" 3. Any Redistribution of source code must retain the above copyright notice
.\"  and the acknowledgment contained in paragraph 6, this list of conditions
.\"  and the disclaimer contained in paragraph 7.
.\" 
.\" 4. Any Redistribution in binary form must reproduce the above copyright
.\"  notice and the acknowledgment contained in paragraph 6, this list of
.\"  conditions and the disclaimer contained in paragraph 7 in the
.\"  documentation and/or other materials provided with the distribution.
.\" 
.\" 5. Redistributions in any form must be accompanied by information on how to
.\"  obtain complete source code for the OpenPBS software and any
.\"  modifications and/or additions to the OpenPBS software. The source code
.\"  must either be included in the distribution or be available for no more
.\"  than the cost of distribution plus a nominal fee, and all modifications
.\"  and additions to the Software must be freely redistributable by any party
.\"  (including Licensor) without restriction.
.\" 
.\" 6. All advertising materials mentioning features or use of the Software must
.\"  display the following acknowledgment:
.\" 
.\"   "This product includes software developed by NASA Ames Research Center,
.\"   Lawrence Livermore National Laboratory, and Veridian Information
.\"   Solutions, Inc.
.\"   Visit www.OpenPBS.org for OpenPBS software support,
.\"   products, and information."
.\" 
.\" 7. DISCLAIMER OF WARRANTY
.\" 
.\" THIS SOFTWARE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT
.\" ARE EXPRESSLY DISCLAIMED.
.\" 
.\" IN NO EVENT SHALL VERIDIAN CORPORATION, ITS AFFILIATED COMPANIES, OR THE
.\" U.S. GOVERNMENT OR ANY OF ITS AGENCIES BE LIABLE FOR ANY DIRECT OR INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
.\" OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" This license will be governed by the laws of the Commonwealth of Virginia,
.\" without reference to its choice of law rules.
.if \n(Pb .ig Iq
.TH qgpumode 1B "" Local PBS
.so ../ers/ers.macros
.Iq
.SH NAME
qgpumode \- change GPU mode
.SH SYNOPSIS
qgpumode -H host -g gpuid -m mode
.SH DESCRIPTION
The 
.B qgpumode 
command will request a MOM to change the mode on one of
it's Nvidia GPUs. The GPU's mode is changed by sending a
.I "GPU Control"
batch request to the batch server. 
.LP
Changing the GPU mode requires PBS Operator or Manager privilege.
It also requires that Torque be configured with --enable-nvidia-gpu.
.LP
.SH OPTIONS
.IP "\-H host" 15
Specifies the host within the cluster on which the GPU is located. The
argument is the name of a host that is a member of the cluster of hosts
managed by the server.
.NL
.IP "\-g gpuid" 15
Specifies the ID of the GPU.
.NL
.IP "\-m mode" 15
Specifies the new mode of the GPU. Valid Nvidia GPU modes are:
.RS
.IP 0 2
- Normal mode.
.IP 1 2
- COMPUTE exclusive thread mode (only one COMPUTE thread is allowed to run on
the GPU).
.IP 2 2
- COMPUTE prohibited mode (no COMPUTE contexts are allowed to run on the GPU).
.IP 3 2
- COMPUTE exclusive process mode (only one COMPUTE process is allowed to run on
the GPU). This mode is new with Nvidia driver version 270.
.NL
.LP
.SH OPERANDS
None
.LP
.Ig
.SH STANDARD ERROR
The qgpumode command will write a diagnostic messages to standard error for each
error occurrence.
.SH EXIT STATUS
Upon successful processing of all the operands presented to the qgpumode command,
the exit status will be a value of zero.
.LP
If the qgpumode command fails to process any operand, the
command exits with a value greater than zero.
.SH SEE ALSO
pbs_mom(8B) and pbs_server(8B)
\" turn off any extra indent left by the Sh macro
.RE


CHANGELOG
c - crash     b - bug fix    e - enhancement    f - new feature  n - note

2.5.8
  b - fix for bugzilla #141 - qsub was overwriting the path variable in PBSD_authenticate
      (backported from 3.0.3)
  b - Reversed the order of $(MOMLIBS) and $(PBS_LIBS) in src/resmom/Makefile.in for 
      LDADD = @PBS_MACH@/libmommach.a $(PBS_LIBS) $(MOMLIBS). This fixes a link 
      problem on Suse Linux. This fixes bugzilla bug 143.


2.5.7
  e - Added new qsub argument -F. This argument takes a quoted string as
      an argument. The string is a list of space separated commandline
      arguments which are available to the job script.
  e - Added an option to asynchronously delete jobs (currently cannot work 
      for qdel -a all due to limitations of single threads) backported from 3.0.2
  c - Fix an issue where job_purge didn't protect key variables that resulted in crashes
  b - fix bugzilla #134, qmgr -= was deleting all entries (backported from 3.0.2)
  b - do not prepend ${HOME} with the current dir for -o and -e in qsub (backported from 3.0.2)
  b - fix jobs named with -J not always having the server name appended correctly 
      (backported from 3.0.2)
  b - make it so that jobs named like arrays via -J have legal output and error file names
      (backported from 3.0.2)
  b - Fixed a bug for high availability. The -l listener option for pbs_server was not
      complete and did not allow pbs_server to properly communicate with the scheduler.
      Also fixed a bug with job dependencies where the second server or later in the
      $TORQUE_HOME/server_name directory was not added as part of the job dependecny
      so dependent jobs would get stuck on hold if the current server was not the first 
      server in the server_name file.
  b - Fixed a potential buffer overflow problem in src/resmom/checkpoint.c function
      mom_checkpoint_recover. I modified the code to change strcpy and strcat to strncpy
      and strncpy.

2.5.6
  b - Made changes to record_jobinfo and supporting functions to be
      able to use dynamically allcated buffers for data. This fixed
      a problem where incoming data overran fixed sized buffers.
  b - restored functionality for -W umask as reported in bugzilla 115
      (backported from 3.0.1)
  b - Updated torque.spec.in to be able to handle the snapshot
      names of builds.
  e - Added new MOM configure option job_starter. This options will execute
      the script submitted in qsub to the executable or script provided
      as the argument to the job_starter option of the MOM configure file.
  b - fix pbs_mom -q to work with parallel jobs (backported from 3.0.1)
  b - fixed a problem with pbs_server high availability where the current 
      server could not keep the HA lock. The problem was a result of truncating
      the directory name where the lock file was kept. TORQUE would fail to 
      validate permissions because it would do a stat on the wrong directory.
  b - Added code to free the mom.lock file during MOM shutdown.
  b - fixed a bug in set_resources that prevented the last resource in a list from being
      checked. As a result the last item in the list would always be added
      without regard to previous entries.
  e - Added new symbol JOB_EXEC_OVERLIMIT. When a job exceeds a limit (i.e. walltime) the
      job will fail with the JOB_EXEC_OVERLIMIT value and
      also produce an abort case for mailing purposes. Previous to this change
      a job exceeding a limit returned 0 on success and no mail
      was sent to the user if requested on abort.
  e - Added options to buildutils/torque.spec.in to conditionally build munge, BLCR, 
      high-availability, cpusets, and spooling. Also allows customization of the
      sendmail path and allows for optional XML conversion to serverdb.
  b - --with-tcp-retry-limit now actually changes things without needing to run autoheader
  e - Added a new queue resource named procct. procct allows the administrator to 
      set queue limits based on the number of total processors requested in a job.
      Patch provided by Martin Siegert.
  e - allow more than 5 concurrent connections to TORQUE using pbsD_connect. Increase it to 10
      (backported from 3.0.1)
  b - fix a segfault when receiving an obit for a job that no longer exists (backported from 3.0.1)
  b - also remove the procct resource when it is applied because of a default (backported from 3.0.1)
  e - allow an administator using the proxy user submission to also set the job id to be used 
      in TORQUE. This makes TORQUE easier to use in grid configurations. (backported from 3.0.2)
  c - fix a segfault when queue has acl_group_enable and acl_group_sloppy set
      true and no acl_groups are defined. (backported from 3.0.1)
  f - Added the ability to detect Nvidia gpus using nvidia-smi (default) or NVML.
      Server receives gpu statuses from pbs_mom. Added server attribute auto_node_gpu
      that allows automatically setting number of gpus for nodes based on gpu
      statuses. Added new configure options --enable-nvidia-gpus,
      --with-nvml-include and --with-nvml-lib.  
  e - The -e and -o options of qsub allow a user to specify a path or optionally a filename for output. 
      If the path given by the user ended with a directory name but no '/' character at the end then
      TORQUE was confused and would not convert the .OU or .ER file to the final output/error file. The
      code has now been changed to stat the path to see if the end path element is a path or directory
      and handled appropriately.
  c - fix a segfault when using --enable-nvidia-gpus and pbs_mom has Nvidia driver
      older than 260 that still has nvidia-smi command
  e - Added new MOM configuration option $rpp_throttle. The syntax for this in the 
      $TORQUE_HOME/mom_priv/config file is $rpp_throttle <value> where value is a long
      representing microseconds. Setting this values causes rpp data to pause after every

src/cmds/qgpumode.c
/*
*         OpenPBS (Portable Batch System) v2.3 Software License
*
* Copyright (c) 1999-2000 Veridian Information Solutions, Inc.
* All rights reserved.
*
* ---------------------------------------------------------------------------
* For a license to use or redistribute the OpenPBS software under conditions
* other than those described below, or to purchase support for this software,
* please contact Veridian Systems, PBS Products Department ("Licensor") at:
*
*    www.OpenPBS.org  +1 650 967-4675                  sales@OpenPBS.org
*                        877 902-4PBS (US toll-free)
* ---------------------------------------------------------------------------
*
* This license covers use of the OpenPBS v2.3 software (the "Software") at
* your site or location, and, for certain users, redistribution of the
* Software to other sites and locations.  Use and redistribution of
* OpenPBS v2.3 in source and binary forms, with or without modification,
* are permitted provided that all of the following conditions are met.
* After December 31, 2001, only conditions 3-6 must be met:
*
* 1. Commercial and/or non-commercial use of the Software is permitted
*    provided a current software registration is on file at www.OpenPBS.org.
*    If use of this software contributes to a publication, product, or
*    service, proper attribution must be given; see www.OpenPBS.org/credit.html
*
* 2. Redistribution in any form is only permitted for non-commercial,
*    non-profit purposes.  There can be no charge for the Software or any
*    software incorporating the Software.  Further, there can be no
*    expectation of revenue generated as a consequence of redistributing
*    the Software.
*
* 3. Any Redistribution of source code must retain the above copyright notice
*    and the acknowledgment contained in paragraph 6, this list of conditions
*    and the disclaimer contained in paragraph 7.
*
* 4. Any Redistribution in binary form must reproduce the above copyright
*    notice and the acknowledgment contained in paragraph 6, this list of
*    conditions and the disclaimer contained in paragraph 7 in the
*    documentation and/or other materials provided with the distribution.
*
* 5. Redistributions in any form must be accompanied by information on how to
*    obtain complete source code for the OpenPBS software and any
*    modifications and/or additions to the OpenPBS software.  The source code
*    must either be included in the distribution or be available for no more
*    than the cost of distribution plus a nominal fee, and all modifications
*    and additions to the Software must be freely redistributable by any party
*    (including Licensor) without restriction.
*
* 6. All advertising materials mentioning features or use of the Software must
*    display the following acknowledgment:
*
*     "This product includes software developed by NASA Ames Research Center,
*     Lawrence Livermore National Laboratory, and Veridian Information
*     Solutions, Inc.
*     Visit www.OpenPBS.org for OpenPBS software support,
*     products, and information."
*
* 7. DISCLAIMER OF WARRANTY
*
* THIS SOFTWARE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. ANY EXPRESS
* OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT
* ARE EXPRESSLY DISCLAIMED.
*
* IN NO EVENT SHALL VERIDIAN CORPORATION, ITS AFFILIATED COMPANIES, OR THE
* U.S. GOVERNMENT OR ANY OF ITS AGENCIES BE LIABLE FOR ANY DIRECT OR INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
* OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
* EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*
* This license will be governed by the laws of the Commonwealth of Virginia,
* without reference to its choice of law rules.
*/
/*
 *
 * qgpumode - Change GPU mode on a MOM
 *
 * Authors:
 *      Al Taufer
 *      Adaptive Computing
 */

#include "cmds.h"
#include <pbs_config.h>   /* the master config generated by configure */

int exitstatus = 0; /* Exit Status */

static void execute(char *, char *, int, char *);

/* qgpumode */

int main(

  int    argc,
  char **argv)

  {
  int   c;
  int   errflg = 0;
  char *gpuid = NULL;
  int   gpumode = -1;

  char *mom_node = NULL;

#define GETOPT_ARGS "H:g:m:"

  while ((c = getopt(argc, argv, GETOPT_ARGS)) != EOF)
    {
    switch (c)
      {

      case 'H':

        if (strlen(optarg) == 0)
          {
          fprintf(stderr, "qgpumode: illegal -H value\n");
          errflg++;
          break;
          }

        mom_node = optarg;

        break;

      case 'g':

        if (strlen(optarg) == 0)
          {
          fprintf(stderr, "qgpumode: illegal -g value\n");
          errflg++;
          break;
          }

        gpuid = optarg;

        break;

      case 'm':

        gpumode = atoi(optarg);
        
        if (gpumode < 0 || gpumode > 3)
          {
          fprintf(stderr, "qgpumode: illegal -m value\n");
          errflg++;
          }

        break;

      default:

        errflg++;

        break;
      }
    }    /* END while (c) */

  if (errflg || (mom_node == NULL) || (gpuid == NULL) || (gpumode == -1))
    {
    static char usage[] = "usage: qgpumode -H host -g gpuid -m mode\n";

    fprintf(stderr, "%s", usage);

    exit(2);
    }

  execute(mom_node, gpuid, gpumode, "");

  exit(exitstatus);

  /*NOTREACHED*/

  return(0);
  }  /* END main() */

/*
 * void execute( char *node, char *gpuid, int mode, char *server )
 *
 * node     The name of the MOM node.
 * gpuid    The id of the GPU
 * mode     The num mode for the GPU
 * server   The name of the server to send to.
 *
 * Returns:
 *  None
 *
 * File Variables:
 *  exitstatus  Set to two if an error occurs.
 */
static void
execute(char *node, char *gpuid, int mode, char *server)
  {
  int ct;         /* Connection to the server */
  int merr;       /* Error return from pbs_manager */
  char *errmsg;   /* Error message from pbs_manager */

  /* The request to change mode */

  if ((ct = cnt2server(server)) > 0)
    {
    merr = pbs_gpumode(ct, node, gpuid, mode);

    if (merr != 0)
      {
      errmsg = pbs_geterrmsg(ct);

      if (errmsg != NULL)
        {
        fprintf(stderr, "qgpumode: %s ", errmsg);
        }
      else
        {
        fprintf(stderr, "qgpumode: Error (%d - %s) changing GPU mode",
                pbs_errno,
                pbs_strerror(pbs_errno));
        }

      if (notNULL(server))
        fprintf(stderr, "@%s", server);

      fprintf(stderr, "\n");

      exitstatus = 2;
      }

    pbs_disconnect(ct);
    }
  else
    {
    fprintf(stderr, "qgpumode: could not connect to server %s (%d) %s\n",
            server,
            pbs_errno,
            pbs_strerror(pbs_errno));
    exitstatus = 2;
    }
  }
